---
title: "CA1012: Abstract types should not have constructors (code analysis)"
description: "Learn about code analysis rule CA1012: Abstract types should not have constructors"
ms.date: 03/11/2019
ms.topic: reference
f1_keywords:
- AbstractTypesShouldNotHaveConstructors
- CA1012
helpviewer_keywords:
- CA1012
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
---
# CA1012: Abstract types should not have constructors

| Item                                     | Value            |
|------------------------------------------|------------------|
| RuleId                                   | CA1012           |
| Category                                 | Microsoft.Design |
| Fix is a breaking or non-breaking change | Non-breaking     |

## Cause

A type is abstract and has a constructor.

By default, this rule only looks at externally visible types, but this is [configurable](#configurability).

## Rule description

Constructors on abstract types can be called only by derived types. Because public constructors create instances of a type and you cannot create instances of an abstract type, an abstract type that has a public constructor is incorrectly designed.

## How to fix violations

To fix a violation of this rule, either make the constructor protected or don't declare the type as abstract.

## When to suppress warnings

Do not suppress a warning from this rule. The abstract type has a public constructor.

## Configurability

You can configure which parts of your codebase to run this rule on, based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:

```ini
dotnet_code_quality.ca1012.api_surface = private, internal
```

You can configure this option for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Configure code analysis rules](../configure-rules.md).

## Example

The following code snippet contains an abstract type that violates this rule.

:::code language="vbnet" source="snippets/vb/all-rules/ca1012-abstract-types-should-not-have-constructors_1.vb" id="snippet1":::

```csharp
// Violates this rule
public abstract class Book
{
    public Book()
    {
    }
}
```

The following code snippet fixes the previous violation by changing the accessibility of the constructor from `public` to `protected`.

```csharp
// Does not violate this rule
public abstract class Book
{
    protected Book()
    {
    }
}
```

:::code language="vbnet" source="snippets/vb/all-rules/ca1012-abstract-types-should-not-have-constructors_1.vb" id="snippet2":::
